### PR TITLE
feat: Add v1 serialization upgrade logic

### LINKED ISSUE #3234: Parameter compatibility of existing widgets between v1 and v2
Here are the parameters that exist in v1 but not v2. The goal of this ticket is transparency as I use some recently written tools to help flesh out the differences between versions:
- [x] src/chart/Area -> sampleData
- [x] src/chart/Bar -> sampleData
- [x] src/chart/Column -> sampleData
- [x] src/chart/Gantt -> sampleData
- [x] src/chart/HexBin -> sampleData
- [x] src/chart/Line -> sampleData
- [ ] src/chart/Pie -> outerText
- [x] src/chart/Scatter -> sampleData
- [x] src/chart/Step -> sampleData
- [x] src/composite/MegaChart -> chart
- [ ] src/composite/MegaChart -> chartType
- [ ] src/composite/MegaChart -> ddlParamsFormat
- [ ] src/composite/MegaChart -> domainAxisTitle
- [ ] src/composite/MegaChart -> infoIcon
- [ ] src/composite/MegaChart -> legendBins
- [ ] src/composite/MegaChart -> legendFormat
- [ ] src/composite/MegaChart -> legendPosition
- [ ] src/composite/MegaChart -> maximizedBackgroundColor
- [ ] src/composite/MegaChart -> showChartSelect
- [ ] src/composite/MegaChart -> showCount
- [ ] src/composite/MegaChart -> showCSV
- [ ] src/composite/MegaChart -> showInfoButton
- [ ] src/composite/MegaChart -> showMaximize
- [ ] src/composite/MegaChart -> showToolbar
- [x] src/composite/MegaChart -> title
- [ ] src/composite/MegaChart -> titleBackgroundColor
- [x] src/composite/MegaChart -> titleFontBold
- [x] src/composite/MegaChart -> titleFontColor
- [x] src/composite/MegaChart -> titleFontFamily
- [x] src/composite/MegaChart -> titleFontSize
- [ ] src/composite/MegaChart -> toolbarShowLegend
- [ ] src/composite/MegaChart -> valueAxisTitle
- [ ] src/form/Slider -> labelRotation
- [ ] src/form/Slider -> logBase
- [ ] src/form/Slider -> loopDiameter
- [ ] src/form/Slider -> loopGutter
- [ ] src/form/Slider -> overlapMode
- [ ] src/form/Slider -> playDiameter
- [ ] src/form/Slider -> playGutter
- [ ] src/form/Slider -> playInterval
- [ ] src/form/Slider -> powExponent
- [ ] src/form/Slider -> showPlay
- [ ] src/form/Slider -> tickFormat
- [ ] src/form/Slider -> type
- [ ] src/graph/Vertex -> icon_shape_diameter
- [x] src/layout/Modal -> bodyPadding
- [x] src/layout/Modal -> headerPadding
- [x] src/layout/Modal -> html
- [ ] src/map/Heat -> meshColor
- [ ] src/map/Heat -> meshStrokeWidth
- [ ] src/other/Legend -> quickUpdateOnClick
- [ ] src/other/NestedTable -> quickUpdateOnClick
- [ ] src/other/Table -> quickUpdateOnClick
- [ ] src/tree/Dendrogram -> zoomToolbar
- [ ] src/tree/Indented -> zoomToolbar

I will check off each of these as they're confirmed to fall within one or more of these categories:
- The marshaller is not dependent upon the widget in question.
- The marshaller is not dependent upon the parameter in question (this only applies to MegaChart and Modal).
- DSP contains no uses of the parameter in question.
- Support for the parameter in question has been added to v2.
- A MegaChart parameter with the same name is available on layout_ChartPanel
- I've added logic to the upgrade script to port the parameter to a viable substitution

### PR DESCRIPTION
**Pull request by jbrundage**
Fixes #3234

Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

## Testing:
Tested within dashy using a global layout JSON (`g_layout`) and DDL1 web service information in the URL.

### CODE REVIEW

**Comment by jbrundage** [comment]
@GordonSmith the upgrade process works without a layout now

And passing a uuid as a dashy URL parameter now does the following:
1) fetches the layout
2) parses the layout's `ddlUrl` for the URL DDL params (ex: Protocol, Hostname, Port)
3) Uses those parsed params as if they'd been provided in the dashy URL

**Comment by jbrundage** [comment]
For testing purposes I used a node script as a proxy to dsp-dev to avoid the CORS limitation, but when dashy is hosted by dsp then they can just directly fetch their layout web service

**Comment by GordonSmith** [comment]
Looks like it needs to be rebased?

File: packages/ddl-shim/src/upgrade.ts [hunk 1]
```diff
@@ -722,7 +728,9 @@ class DDLUpgrade {
     }
 }
 
-export function upgrade(ddl: DDL1.IDDL, baseUrl?: string, wuid?: string, toLowerCase: boolean = true): DDL2.Schema {
+export function upgrade(ddl: DDL1.IDDL, baseUrl?: string, wuid?: string, toLowerCase: boolean = true, dermatologyJson = {}): DDL2.Schema {
```
Comment (GordonSmith) [inline packages/ddl-shim/src/upgrade.ts]: Change to `dermatologyJson?: object`

File: packages/ddl-shim/src/upgrade.ts [hunk 2]
```diff
@@ -722,7 +728,9 @@ class DDLUpgrade {
...
     const ddlUp = new DDLUpgrade(ddl, baseUrl, wuid, toLowerCase);
-    return ddlUp.write();
+    const ddl2 = ddlUp.write();
+    dermatologyUpgrade(ddl2, dermatologyJson);
```
Comment (GordonSmith) [inline packages/ddl-shim/src/upgrade.ts]: Only call if dermatologyJson is not undefined

File: packages/layout/src/Border2.ts [hunk 3]
```diff
@@ -190,6 +195,7 @@ export interface Border2 {
 }
 Border2.prototype.publish("top", null, "widget", "Top Widget", undefined, { render: false });
 Border2.prototype.publish("topOverlay", false, "boolean", "Overlay Top Widget");
+Border2.prototype.publish("topHeight", null, "number", "Top Fixed Height (pixels)", undefined, { optional: true });
```
Comment (GordonSmith) [inline packages/layout/src/Border2.ts]: Should be separate PR

File: packages/layout/src/ChartPanel.ts [hunk 4]
```diff
@@ -517,7 +522,11 @@ ChartPanel.prototype.publishProxy("titleIcon", "_titleBar");
 ChartPanel.prototype.publishProxy("titleIconFont", "_titleBar");
 ChartPanel.prototype.publishProxy("titleFont", "_titleBar");
 ChartPanel.prototype.publishProxy("titleIconFontSize", "_titleBar");
+ChartPanel.prototype.publishProxy("titleIconFontColor", "_titleBar");
+ChartPanel.prototype.publishProxy("titleFontBold", "_titleBar");
 ChartPanel.prototype.publishProxy("titleFontSize", "_titleBar");
+ChartPanel.prototype.publishProxy("titleFontColor", "_titleBar");
+ChartPanel.prototype.publishProxy("titleBackgroundColor", "_titleBar");
```
Comment (GordonSmith) [inline packages/layout/src/ChartPanel.ts]: should be separate PR

File: packages/layout/src/Grid.ts [hunk 5]
```diff
@@ -515,6 +555,8 @@ export class Grid extends HTMLWidget {
 Grid.prototype._class += " layout_Grid";
 
 Grid.prototype.publish("designMode", false, "boolean", "Design Mode", null, { tags: ["Basic"] });
+Grid.prototype.publish("adaptiveSnapping", false, "boolean", "Adapt the snapping columns and rows to match the content", null, { tags: ["Basic"], disable: w => !w.designMode() });
+Grid.prototype.publish("showLanes", true, "boolean", "Show snapping lanes when in design mode", null, { tags: ["Basic"], disable: w => !w.designMode() });
```
Comment (GordonSmith) [inline packages/layout/src/Grid.ts]: Should be separate PR

File: packages/layout/src/Surface.ts [hunk 6]
```diff
@@ -153,6 +156,7 @@ export class Surface extends HTMLWidget {
 Surface.prototype._class += " layout_Surface";
 
 Surface.prototype.publish("title", "", "string", "Title", null, { tags: ["Intermediate"] });
+Surface.prototype.publish("altText", null, "string", "Alt text", null, { optional: true });
```
Comment (GordonSmith) [inline packages/layout/src/Surface.ts]: Should be separate PR (can probably be same one as Grid

File: packages/marshaller/src/dashy.ts [hunk 7]
```diff
@@ -144,8 +144,8 @@ export class Dashy extends SplitPanel {
         });
     }
 
-    importDDL(ddl: DDL1.DDLSchema | DDL2.Schema, baseUrl?: string, wuid?: string): Promise<void> {
-        const ddl2: DDL2.Schema = isDDL2Schema(ddl) ? ddl : upgrade(ddl, baseUrl, wuid);
+    importDDL(ddl: DDL1.DDLSchema | DDL2.Schema, baseUrl?: string, wuid?: string, dermatologyJson?: any): Promise<void> {
```
Comment (GordonSmith) [inline packages/marshaller/src/dashy.ts]: Chaange to `dermatologyJson?: object`

File: packages/marshaller/src/ddl2/dashboard.ts [hunk 8]
```diff
@@ -198,8 +199,27 @@ class DashboardDockPanel extends DockPanel implements IClosable, IVizPopupPanelO
...
+
+    private _designModeButton = new Button().faChar("fa-object-group").tooltip("Design Mode...")
+        .on("click", () => {
+            if (this._dockPanel && (this._dockPanel as any).designMode) {
+                (this._dockPanel as any)
+                    .designMode(!(this._dockPanel as any).designMode())
+                    .render()
+                    ;
+            }
+        });
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/dashboard.ts]: Design mode should be hidden when not a Grid dashboard

File: packages/marshaller/src/ddl2/dashboard.ts [hunk 9]
```diff
@@ -321,7 +341,7 @@ export class Dashboard extends ChartPanel {
             const dashboard = new Dashboard(ec)
                 .target(target)
                 .hideSingleTabs(true)
-                .titleVisible(false)
+                .titleVisible(true)
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/dashboard.ts]: I think this is meant to be `false`?

File: packages/marshaller/src/ddl2/dashboard.ts [hunk 10]
```diff
@@ -342,8 +363,8 @@ export class Dashboard extends ChartPanel {
             })
             ;
         this
-            .buttons([this._addButton, this._removeButton, new Spacer(), this._reloadButton, new Spacer(), this._addSamples])
-            .widget(this._dockPanel)
+            .buttons([this._designModeButton, new Spacer(), this._addButton, this._removeButton, new Spacer(), this._reloadButton, new Spacer(), this._addSamples])
+            .widget(this._dockPanel as any)
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/dashboard.ts]: A little ugly, but probably unavoidable.

File: packages/marshaller/src/ddl2/dashboard.ts [hunk 11]
```diff
@@ -383,6 +404,15 @@ export class Dashboard extends ChartPanel {
 
     // _delayLayout;
     restore(_: DDL2.Schema): this {
+        if (_ && _.properties && _.properties.layout instanceof Array) {
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/dashboard.ts]: Would be nicer if this test was more explicit (may not be possible as it is)

File: packages/marshaller/src/ddl2/dashboard.ts [hunk 12]
```diff
@@ -383,6 +404,15 @@ export class Dashboard extends ChartPanel {
...
+            this.widget().target(null);
+            this._dockPanel = new DashboardGrid(this.elementContainer())
+                .on("vizActivation", (elem: Element) => {
+                    this.vizActivation(elem);
+                })
+                ;
+            this.widget(this._dockPanel as any);
+        }
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/dashboard.ts]: Needs an `else` statement to create the `DashboardDockPanel`

File: packages/marshaller/src/ddl2/grid.ts [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+        // this.fitTo_default("width");
+        this.designMode_default(false);
+        this.adaptiveSnapping_default(true);
+        this.showLanes_default(false);
+        this.surfacePadding_default("0");
+        this.surfaceBorderWidth_default(0);
+        // this.snappingColumns_default(900);
+        // this.snappingRows_default(900);
+        // this.squareDesignMode_default(true);
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/grid.ts]: remove commented code

File: packages/marshaller/src/ddl2/grid.ts [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    private _popups: VizPopupPanel[] = [];
+    private _popupIdx: {
+        [id: string]: {
+            panel: VizPopupPanel,
+            monitorHandle: IMonitorHandle
+        }
+    } = {};
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/grid.ts]: Should probably become a "PopupManager" class so that code can be commoned up between the two dashboard types?

File: packages/marshaller/src/ddl2/grid.ts [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+
+    updateTitle(w: Widget) {
+        const cell: Cell = this.getWidgetCell(w.id());
+        const element: Element = this._ec.element(w);
+        // const errors = element.validate();
+        const tabTitle = this.tabTitle(element);
+        // const caption = errors.map(err => `${err.source}:  ${err.msg}`).join("\n");
+        (cell.widget() as any)
+            .title(tabTitle)
+            // .altText(caption)
```
Comment (GordonSmith) [inline packages/marshaller/src/ddl2/grid.ts]: remove commented code

File: demos/dashy/index.html [hunk 16]
```diff
@@ -36,6 +36,8 @@
     <script>
         var hpccLoader = window["@hpcc-js/loader"];
         var vizLib = window["vizLib"];
+        window.g_layout = {};
+        window.g_uuid_url = "http://localhost:8091/";
```
[the remark this answers is not in the document]
Comment (jbrundage) [reply demos/dashy/index.html]: The UUID url param doesn't work without this: `g_uuid_url`

A proxy needs to be set up that points to whichever dsp environment the layout info needs to be fetched from. The `g_uuid_url` global allows the dsp folks to set this  base url. Should set it up as a proper `App` class attribute?

File: packages/marshaller/src/ddl2/grid.ts (hunk 14 above)
Comment (jbrundage) [reply packages/marshaller/src/ddl2/grid.ts]: Can we open an issue for this and handle it separately?

Comment (GordonSmith) [reply packages/marshaller/src/ddl2/grid.ts]: Sure

**Comment by GordonSmith** [comment]
@jbrundage can you squash to a single commit for pulling (and open the new Jira ticket as mentioned)?

File: packages/ddl-shim/src/dermatology.ts [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+            ]
+        };
+        ddl2.properties.layout.push(cellPosition);
+    });
+}
+
+function findKeyVal(object: any, key: any, val: any) {
+    let value;
+
+    Object.keys(object).some(k => {
```
[the remark this answers is not in the document]
Comment (jbrundage) [reply packages/ddl-shim/src/dermatology.ts]: The loop iterates the object keys searching for a key+value pair that matches the given key and value. I used `some` to be able to escape the loop after the first match is found.

[the remark this answers is not in the document]
Comment (GordonSmith) [reply packages/ddl-shim/src/dermatology.ts]: Without the loop wouldn't the following do the same:
```
if (object[k] && object[k] === val) {...}
```
